### ISSUE TITLE
Error message on client and server

### ISSUE THREAD

**Issue by uncle-yura** [mbuesch/awlsim#22]
Is it possible to do precompilation on the client side? A commonplace error in the name causes a stop CPU and there are no error messages on the client side.
```
awlsim-linuxcnc: [133.250] Awlsim error:
  Multiple definitions of FC 1.
```
When you load an STL file, the error is at least highlighted in red, but when you work in FBD there are no messages or highlighting.
Also, the FBD source file is not indicated in the error, which greatly complicates the search.
```
awlsim-linuxcnc: [1919.813] Awlsim error:
  Unconnected pin "Q" found in FUP element XOR(x3/y10, 2c096344..65166).
The reporting FUP/FBD element is:
FUP-grid
```
If there is no way to precompile on the client, is there a way to display the error in the server GUI?

**Comment by mbuesch** [comment]
Thanks for your comment.

These are known issues and I agree that these should be fixed.
But let me comment on each of them:

> Is it possible to do precompilation on the client side?

Yes. We actually do that already.
Compilation (at client side it's called validation) is attempted after each source change (life in the background).
On a best effort basis an attempt is being made to map the resulting errors to the source being edited. If there's a simple typo in an STL command, it will be highlighted.
However, mapping these errors to the root cause of the problem is not always easy. So it often fails.

But it would possibly be a good idea to (have an option to) trigger validation on the client side before download and then prevent the download of the program, if validation fails in any way.
I'll track this enhancement in the TODO list.

> Multiple definitions of FC 1.

Basic errors like clashing FC names are especially hard to map to the root cause.
Yes, that's a weak point in awlsim's compiler.

> but when you work in FBD there are no messages or highlighting.

Yes, that is known.
There is no graphical mapping of the AWL or FUP compiler error back to the FUP/FBD diagram.
However in the error message text there usually are the coordinates of the FUP element that caused the problem (or is close to the root cause).
In your example that would be x3/y10.

However a graphical marking of errors would be preferred. I'll track this in the TODO list.

> is there a way to display the error in the server GUI?

If there is a compilation error or any other error on the server and a client is connected, then the server-side error is displayed on the connected client. It appears to the user just like any other error on the usual AwlsimError message box.

So that should actually work right now. Do you have an example of where it doesn't work?

**Comment by uncle-yura** [comment]
> If there is a compilation error or any other error on the server and a client is connected, then the server-side error is displayed on the connected client. It appears to the user just like any other error on the usual AwlsimError message box.
> So that should actually work right now. Do you have an example of where it doesn't work?

As I wrote above, there are no errors when editing FBD.

What I did:
1. Created a new circuit named "test"
2. Gave the wrong address "FC 1" and added a block with unconnected inputs / outputs
3. Saved
4. I downloaded all sources to CPU

What I received from the client:
1. Server response timeout message box

What I got from the server side:
1. The console message about an unconnected pin

Client and server on the same machine, this is a fresh image of LinuxCNC, Debian Stretch, kernel 4.9.0-8.

**Comment by mbuesch** [comment]
> As I wrote above, there are no errors when editing FBD.

Ok sorry, that was a misunderstanding on my side.
There currently is no life-validation of FUP/FBD.
So, yes. A server side error is expected here. That might be mitigated by a pre-download validation.

>What I received from the client:
>Server response timeout message box

Ok, that should never happen. I think the server might have crashed.
Do you have the full server side log?
An awlpro file for me to reproduce the bug would be helpful, too.

Thanks for your help.

**Comment by mbuesch** [comment]
Ok, I see now.
This is a problem in the LinuxCNC module.

I guess you have already noticed that the Awlsim LinuxCNC module is very experimental and hardly tested at all. :)
Sorry for that.
I hope that doesn't cause too much inconvenience.
Your bug reports help to fix the module. Please keep them coming.

**Comment by uncle-yura** [comment]
I’m very interested in your program, because I don’t like the functionality and convenience of classicladder in LinuxCNC. Unfortunately, my knowledge is not yet enough to help with anything other than ideas and testing.

**Comment by mbuesch** [comment]
> Unfortunately, my knowledge is not yet enough to help with anything other than ideas and testing.

Your ideas and testing is more than appreciated. Thanks.

**Comment by mbuesch** [comment]
The problem with the forwarding of the error messages from the LinuxCNC module to the Awlsim GUI has been fixed in: 7bd1ae190e7d4db296499bbaf137a41f290b2043
